### PR TITLE
Enable the CANFD bit back on len member

### PR DESCRIPTION
**Pull request by asconcepcion**
If a write operation of a CANFD frame doesn't succeed, the CANFD bit has to be
enabled back on len before inserting the frame on the buffer, so that the next
time it tries to send the frame, this gets properly decoded.

### CODE REVIEW

**Comment by hartkopp** [comment]
Does it make sense to put this CANFD_FRAME flag into the 'len' field?
I would suggest to define e.g. a `#define CANFD_FDF 0x04` bit in the 'flags' field instead.

**Comment by asconcepcion** [comment]
> Does it make sense to put this CANFD_FRAME flag into the 'len' field?
> I would suggest to define e.g. a `#define CANFD_FDF 0x04` bit in the 'flags' field instead.

My understanding is that frame can actually be either a CANFD or a basic CAN packet, but the latter one doesn't have the 'flags' field inside, just the message ID, len and data.

**Comment by hartkopp** [comment]
You are using the struct canfd_frame anyway, that's why you can access the 'len' element. In struct can_frame there is no 'len' element but 'can_dlc'. A potential 0x04 bit in the 'flags' element is not used inside the kernel (which distinguishes CAN from CAN FD via the can(fd)_frame structure length). So when you need this bit to decide (in user space) whether the struct canfd_frame contains a CAN frame -or- a CAN FD frame, I would suggest to use the 'flags' element.
There's no need to do bit masking on the 'len' field then.

**Comment by asconcepcion** [comment]
I see your point... I am not very familiar with the whole code, but that would require to change the way in which the basic CAN frames are "serialized" in a cannelloni packet to be forwarded via Ethernet. As it is now, the different CAN frames are serialized like this (see [here](https://github.com/mguentner/cannelloni/blob/master/parser.cpp#L96-L107)):

- For CANFD: it serializes (sends) the message id, len (length of data), flags and data.
- For basic CAN: it serializes (sends) the message id, can_dlc (length of data), and data.

In the piece of code I'm modifying here, it's true it is using a pointer to a canfd_frame struct, but as it is only dereferencing the "len' field (which is a member that has the same size, offset and meaning as the 'can_dlc' element of a can_frame struct), this can be used also for a can_frame struct.

In short, as I see it, to do what you propose we would need to add also a flag field when encapsulating a "can_frame" in  the Ethernet packet (i.e. just sending canfd_frames via ethernet, and then convert them to can_frame if needed).

**Comment by mguentner** [comment]
@asconcepcion Good catch! Your observation and fix is valid.
All parts that want to get the real length of a can frame use the `canfd_len` helper function as defined here: https://github.com/mguentner/cannelloni/blob/master/cannelloni.h#L74

I like the idea @hartkopp suggested to set `0x04` in `frame.flags` in the deserializer. 
To keep the userspace -> kernel interface clean it makes sense to clear the `0x04` before handing it to the kernel and setting it again if the transmit failed like this PR implements for `frame.len`. Otherwise we gamble that the kernel does not check the flags for wrong values and never will.

@asconcepcion Would you be willing to update this PR accordingly? I can assist if you have questions.

**Comment by mguentner** [comment]
Addition: 0x04 set in `canfd.flags` needs to be cleared also when it is transmitted via network in its serialized form. Here the signalling bit should remain 0x80 in `len` to maintain backwards compatibility.

**Comment by asconcepcion** [comment]
Thanks @mguentner =) ! 
I can try to update this PR accordingly, but not sure yet I got the proposal. So, the idea would be to keep using the MSB of `canfd_frame.len`/`can_frame.can_dlc` fields when serializing the CAN frames on a cannelloni packet as until now, right? If so, what would be the purpose of 0x04 in `canfd_frame.flags`?

**Comment by mguentner** [comment]
My proposal was to use `frame->flags` internally, within cannelloni, to distinguish whether a frame is **FD** or not instead of `frame->len`. The wire protocol & packet format must not change as there is no dire need for it.

So basically replacing all parts where `frame>len` is checked for a set MSB.

However I just quickly looked through the code and concluded that this does not bring a real benefit as the logic stays the same just the bit would move from `frame->len` to `frame->flags`.

I suggest to simply merge this PR as it is.

**Comment by asconcepcion** [comment]
Yes, I agree... I think that using `frame->flags` within cannelloni would have been OK if the can_frame struct also had had it, but as it is not the case, we would need to artificially add a basic CAN frame when serializing it on a cannelloni packet. Using the MSB of `frame->len`/`frame->can_dlc` should be safe since with the maximum data length for both frames shouldn't reach that bit.
